### ISSUE TITLE
spec.input_namespace('pseudos') always creates a key 'pseudos' in self.inputs 

### ISSUE THREAD

**Issue by bosonie** [aiidateam/aiida-core#3749]
While playing around with the siesta plugin, I noticed this behaviour.
Defining:
spec.input_namespace('pseudos', required="a", ..) 
always creates a key 'pseudos' in self.inputs, independently if a=false or a=true
Here, for instance, the explicit string I get when I do print(self.inputs):

<AttributesFrozendict {'code': <Code: Remote code 'Siesta-4.0.2' on kay, pk: 2, uuid: 32a019cd-1ca9-47ac-a3f0-66bdc4d26726>, 'structure': <StructureData: uuid: 93e4dac9-e991-4442-b00c-8e87a2d71f63 (pk: 884)>, 'parameters': <Dict: uuid: b02f5e93-5cc1-4135-bee9-aeca0d5867f8 (pk: 885)>, 'metadata': <AttributesFrozendict {'label': 'TestOnSiliconBulk', 'options': <AttributesFrozendict {'max_wallclock_seconds': 360, 'resources': {'num_machines': 1, 'num_mpiprocs_per_machine': 1, 'default_mpiprocs_per_machine': 20}, 'input_filename': 'aiida.fdf', 'output_filename': 'aiida.out', 'scheduler_stdout': '_scheduler-stdout.txt', 'scheduler_stderr': '_scheduler-stderr.txt', 'custom_scheduler_commands': '', 'withmpi': False, 'mpirun_extra_params': [], 'import_sys_environment': True, 'environment_variables': {}, 'prepend_text': '', 'append_text': '', 'parser_name': 'siesta.parser'}>, 'store_provenance': True, 'call_link_label': 'CALL', 'dry_run': False}>, 'pseudos': <AttributesFrozendict {}>}>

You see the last entry is 'pseudos', that I haven't specified.
I don't know if this is a wanted feature, but for sure it is confusing for many developers as I see in the plugins lines like this one:
https://github.com/aiidateam/aiida-quantumespresso/blob/01ee430a550db789ad9c91a323408f35568e40d5/aiida_quantumespresso/utils/pseudopotential.py#L35
that are useless since pseudos is never None.

Personally I think that at least with required="false", the 'pseudos' key shouldn't be there, but maybe you have arguments against it.
Also, I think the role of required="" for input_namespace is not so clear.

**Comment by sphuber** [comment]
I agree that the behavior can be confusing and one might expect it not to be there when not explicitly passed. The code that concerns this behavior is actually part of `plumpy` so I will move it there I think for further discussion.

The line in the `validate_and_prepare_pseudos_inputs` function from `aiida-quantumespresso` plugin package that you link is actually not a problem I think. When considering the function in isolation the behavior is correct. Even if you pass an empty dictionary for a dictionary, it will be caught later i n line 44 where it will check that all pseudos are present.

I will see if I can change the code and make a PR with a fix and see what the consequences are.
